metaseq/distributed/rendezvous.py

```diff
@@ -70,4 +70,94 @@ def _error(msg):
 
 
+STORE_BASED_BARRIER_MARKER = "store_based_barrier_key:"
+
+
+class ComplicitStore(Store):
+    def __init__(self, store: Store, world_size: int):
+        super().__init__()
+        self.store = store
+        self.world_size = world_size
+
+    def set(self, *args, **kwargs):
+        return self.store.set(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        return self.store.get(*args, **kwargs)
+
+    def add(self, key: str, *args, **kwargs):
+        # The marker isn't always a prefix: it's sometimes prefixed with default_pg/.
+        if STORE_BASED_BARRIER_MARKER in key:
+            return self.world_size
+        return self.store.add(key, *args, **kwargs)
+
+    def compare_set(self, *args, **kwargs):
+        return self.store.compare_set(*args, **kwargs)
+
+    def delete_key(self, *args, **kwargs):
+        return self.store.delete_key(*args, **kwargs)
+
+    def num_keys(self, *args, **kwargs):
+        return self.store.num_keys(*args, **kwargs)
+
+    def set_timeout(self, *args, **kwargs):
+        return self.store.set_timeout(*args, **kwargs)
+
+    def wait(self, *args, **kwargs):
+        return self.store.wait(*args, **kwargs)
+
+
+STORE = None
+
+
+def meet_at_rendezvous(url, **kwargs):
+    assert url == "barrierlessenv://"
+    rendezvous_iterator = rendezvous("env://")
+    store, rank, world_size = next(rendezvous_iterator)
+    global STORE
+    STORE = ComplicitStore(store, world_size)
+    return iter([(STORE, rank, world_size)])
+
+
+def _tcp_retry_barrierless_rendezvous_handler(url: str, timeout=default_pg_timeout, **kwargs):
+    def _error(msg):
+        return ValueError("barrierlesstcpr:// rendezvous: " + msg)
+
+    global STORE
+
+    result = urlparse(url)
+    if not result.port:
+        raise _error("port number missing")
+    query_dict = _query_to_dict(result.query)
+    if "rank" not in query_dict:
+        raise _error("rank parameter missing")
+    if "world_size" not in query_dict:
+        raise _error("world size parameter missing")
+
+    rank = int(query_dict["rank"])
+    world_size = int(query_dict["world_size"])
+    assert result.hostname is not None
+
+    for tries in range(1, RETRIES + 1):
+        try:
+            store = _create_c10d_store(
+                result.hostname, result.port, rank, world_size, timeout
+            )
+            logger.warning(
+                f"Successfully connected to primary node on attempt {tries}/{RETRIES}"
+            )
+            STORE = ComplicitStore(store, world_size)
+            yield (STORE, rank, world_size)
+            break
+        except RuntimeError as re:
+            logger.error(
+                f"Failed to connect to primary node on attempt {tries}/{RETRIES}: {re}"
+            )
+            time.sleep(COOLDOWN * (2**tries))
+
+    # If this configuration is invalidated, there is nothing we can do about it
+    raise RuntimeError("Unable to perform re-rendezvous using barrierlesstcpr:// method")
```

Comment:
I think this can be simplified substantially to something like this:
```suggestion
    assert url.startswith("barrierlesstcpr://")
    url = url.replace("barrierlesstcpr://", "tcpr://")
    rendezvous_iterator = rendezvous(url)
    store, rank, world_size = next(rendezvous_iterator)
    global STORE
    STORE = ComplicitStore(store, world_size)
    return iter([(STORE, rank, world_size)])
```

Reply:
And, in fact, it could even further be merged with the `meet_at_rendezvous` function above, to deduplicate the two almost-identical functions.